### PR TITLE
Add tags to compare API

### PR DESCRIPTION
**Pull request by jonkeane**
This might be a super hacky implementation, but it seems to work in the tests at least. This info is (mostly) repeated in the benchmark name now, but I'm hoping to be able to pull something from the API and then parse the tags themselves for creating a diff report kind of like those I [made when helping Romain](https://github.com/apache/arrow/pull/9615#issuecomment-841253893)

### CODE REVIEW

**Comment by jonkeane** [comment]
I can add the appropriate tests to not have the coverage drop — it shouldn't be too hard to construct test cases for them (though some will be more contrived than others i.e. getting the tags from the contender when the baseline doesn't have any)

**Comment by dianaclarke** [comment]
This looks great, thanks! Rather than complicate the code and API, just go ahead and always include tags.

File: conbench/api/_comparator.py [hunk 1]
```diff
@@ -15,21 +15,23 @@ def change_fmt(value):
 
 
 class BenchmarkResult:
-    def __init__(self, id, batch_id, run_id, unit, value, batch, benchmark):
+    def __init__(self, id, batch_id, run_id, unit, value, batch, benchmark, tags=None):
```
Comment (dianaclarke) [inline conbench/api/_comparator.py]: Simplify by always including tags.

File: conbench/api/_comparator.py [hunk 2]
```diff
@@ -15,21 +15,23 @@ def change_fmt(value):
...
         self.unit = unit
         self.batch = batch
         self.benchmark = benchmark
         self.value = decimal.Decimal(value)
+        self.tags = tags
 
 
 class BenchmarkComparator:
-    def __init__(self, baseline, contender, threshold=None):
+    def __init__(self, baseline, contender, threshold=None, include_tags=False):
```
Comment (dianaclarke) [inline conbench/api/_comparator.py]: Simplify by always including tags.

File: conbench/api/_comparator.py [hunk 3]
```diff
@@ -15,21 +15,23 @@ def change_fmt(value):
...
         self.baseline = BenchmarkResult(**baseline) if baseline else None
         self.contender = BenchmarkResult(**contender) if contender else None
         self.threshold = threshold if threshold is not None else THRESHOLD
+        self.include_tags = include_tags
```
Comment (dianaclarke) [inline conbench/api/_comparator.py]: Simplify by always including tags.

File: conbench/api/_comparator.py [hunk 4]
```diff
@@ -108,6 +118,7 @@ def formatted(self):
             "contender_run_id": self.contender.run_id if self.contender else None,
             "unit": self.unit,
             "less_is_better": self.less_is_better,
+            **({"tags": self.tags} if self.include_tags else {}),
```
Comment (dianaclarke) [inline conbench/api/_comparator.py]: Simplify by always including tags.

File: conbench/api/_comparator.py [hunk 5]
```diff
@@ -129,20 +140,26 @@ def compare(self):
...
         }
 
 
 class BenchmarkListComparator:
-    def __init__(self, pairs, threshold=None):
+    def __init__(self, pairs, threshold=None, include_tags=False):
         self.pairs = pairs
         self.threshold = threshold if threshold is not None else THRESHOLD
+        self.include_tags = include_tags
```
Comment (dianaclarke) [inline conbench/api/_comparator.py]: Simplify by always including tags.

File: conbench/api/_examples.py [hunk 6]
```diff
@@ -102,13 +102,7 @@ def _api_commit_entity(commit_id):
...
 
 
-def _api_compare_entity(
-    benchmark_ids,
-    batch_ids,
-    run_ids,
-    batch,
-    benchmark,
-):
+def _api_compare_entity(benchmark_ids, batch_ids, run_ids, batch, benchmark, tags=None):
```
Comment (dianaclarke) [inline conbench/api/_examples.py]: Simplify by always including tags.

File: conbench/api/_examples.py [hunk 7]
```diff
@@ -135,6 +130,7 @@ def _api_compare_list(
     run_ids,
     batches,
     benchmarks,
+    tags=None,
```
Comment (dianaclarke) [inline conbench/api/_examples.py]: Simplify by always including tags.

File: conbench/api/compare.py [hunk 8]
```diff
@@ -55,6 +60,7 @@ def get(self, compare_ids):
           - Compare
         """
         raw = f.request.args.get("raw", "false").lower() in ["true", "1"]
+        tags = f.request.args.get("tags", "false").lower() in ["true", "1"]
```
Comment (dianaclarke) [inline conbench/api/compare.py]: Simplify by always including tags. Remove this query param.

File: conbench/api/compare.py [hunk 9]
```diff
@@ -116,6 +132,7 @@ def get(self, compare_ids):
...
```
Comment (dianaclarke) [inline conbench/api/compare.py]: Simplify by always including tags. Remove this query param.

File: conbench/api/_examples.py [hunk 10]
```diff
@@ -270,6 +268,14 @@ def _api_run_entity(run_id, commit_id, machine_id, now, baseline_id):
     ["some-run-uuid-1", "some-run-uuid-2"],
     "file-read",
     "snappy, nyctaxi_sample, parquet, arrow",
+    {
+        "compression": "snappy",
+        "cpu_count": 2,
+        "dataset": "nyctaxi_sample",
+        "file_type": "parquet",
+        "input_type": "arrow",
+        "name": "read",
```
Comment (dianaclarke) [inline conbench/api/_examples.py]: `"name": "read",`

I think this should be changed to `file-read`.

File: conbench/api/_examples.py [hunk 11]
```diff
@@ -281,6 +287,24 @@ def _api_run_entity(run_id, commit_id, machine_id, now, baseline_id):
...
         "snappy, nyctaxi_sample, parquet, arrow",
     ],
+    [
+        {
+            "compression": "snappy",
+            "cpu_count": 2,
+            "dataset": "nyctaxi_sample",
+            "file_type": "parquet",
+            "input_type": "arrow",
+            "name": "read",
```
Comment (dianaclarke) [inline conbench/api/_examples.py]: `"name": "read",`

I think this should be changed to `file-read`.

File: conbench/api/_examples.py [hunk 12]
```diff
@@ -281,6 +287,24 @@ def _api_run_entity(run_id, commit_id, machine_id, now, baseline_id):
...
+        },
+        {
+            "compression": "snappy",
+            "cpu_count": 2,
+            "dataset": "nyctaxi_sample",
+            "file_type": "parquet",
+            "input_type": "arrow",
+            "name": "write",
```
Comment (dianaclarke) [inline conbench/api/_examples.py]: `"name": "read",`

I think this should be changed to `file-write`.

File: conbench/tests/api/test_compare.py [hunk 13]
```diff
@@ -59,7 +59,16 @@ def test_compare(self, client):
...
+            tags={
+                "dataset": "nyctaxi_sample",
+                "cpu_count": 2,
+                "file_type": "parquet",
+                "input_type": "arrow",
+                "compression": "snappy",
+                "name": "read",
+            },
         )
+        # import pdb; pdb.set_trace()
```
Comment (dianaclarke) [inline conbench/tests/api/test_compare.py]: Remove pdb

File: conbench/tests/api/test_compare.py [hunk 14]
```diff
@@ -59,7 +59,16 @@ def test_compare(self, client):
             run_ids,
             "read",
             CASE,
+            tags={
+                "dataset": "nyctaxi_sample",
+                "cpu_count": 2,
+                "file_type": "parquet",
+                "input_type": "arrow",
+                "compression": "snappy",
+                "name": "read",
```
Comment (dianaclarke) [inline conbench/tests/api/test_compare.py:68]: Unlike the others, this `"read"` is correct.

File: conbench/tests/api/test_compare.py (hunk 13 above)
Comment (jonkeane) [reply conbench/tests/api/test_compare.py]: Ack, I thought I had caught all of these already!

File: conbench/hacks.py [hunk 15]
```diff
@@ -46,6 +46,15 @@ def set_display_batch(benchmark):
         benchmark.display_batch = batch
 
 
+def set_tags(benchmark):
```
Comment (dianaclarke) [inline conbench/hacks.py]: This isn't needed.

If you apply this diff, I think you'll find all tests pass.

```
diff --git a/conbench/api/compare.py b/conbench/api/compare.py
index 9fd4bbf..b5775e8 100644
--- a/conbench/api/compare.py
+++ b/conbench/api/compare.py
@@ -6,7 +6,7 @@ from ..api._comparator import BenchmarkComparator, BenchmarkListComparator
 from ..api._endpoint import ApiEndpoint
 from ..entities._entity import NotFound
 from ..entities.summary import Summary
-from ..hacks import set_display_batch, set_display_name, set_tags
+from ..hacks import set_display_batch, set_display_name
 
 
 def _compare_entity(summary):
@@ -18,7 +18,7 @@ def _compare_entity(summary):
         "unit": summary.unit,
         "benchmark": summary.display_name,
         "batch": summary.display_batch,
-        "tags": summary.tags,
+        "tags": summary.case.tags,
     }
 
 
@@ -71,8 +71,6 @@ class CompareBenchmarksAPI(ApiEndpoint):
         set_display_name(contender_summary)
         set_display_batch(baseline_summary)
         set_display_batch(contender_summary)
-        set_tags(baseline_summary)
-        set_tags(contender_summary)
 
         baseline = _compare_entity(baseline_summary)
         contender = _compare_entity(contender_summary)
@@ -143,12 +141,10 @@ class CompareBatchesAPI(ApiEndpoint):
         for summary in baselines:
             set_display_name(summary)
             set_display_batch(summary)
-            set_tags(summary)
             self._add_pair(pairs, summary, "baseline")
         for summary in contenders:
             set_display_name(summary)
             set_display_batch(summary)
-            set_tags(summary)
             self._add_pair(pairs, summary, "contender")
 
         if raw:
diff --git a/conbench/hacks.py b/conbench/hacks.py
index e536d1e..7b72b99 100644
--- a/conbench/hacks.py
+++ b/conbench/hacks.py
@@ -46,15 +46,6 @@ def set_display_batch(benchmark):
         benchmark.display_batch = batch
 
 
-def set_tags(benchmark):
-    is_api = isinstance(benchmark, dict)
-    tags = benchmark["tags"] if is_api else benchmark.case.tags
-    if is_api:
-        benchmark["tags"] = tags
-    else:
-        benchmark.tags = tags
-
-
 def sorted_data(benchmarks):
     data = []
     for benchmark in benchmarks:
```

File: conbench/api/_comparator.py [hunk 16]
```diff
@@ -140,9 +151,17 @@ def __init__(self, pairs, threshold=None):
     def formatted(self):
         for pair in self.pairs.values():
             baseline, contender = pair.get("baseline"), pair.get("contender")
-            yield BenchmarkComparator(baseline, contender, self.threshold).formatted()
+            yield BenchmarkComparator(
+                baseline,
+                contender,
+                self.threshold,
```
Comment (dianaclarke) [inline conbench/api/_comparator.py]: If you remove the trailing comma after `self.threshold,` there won't be a diff here. (after you re-black it).

File: conbench/api/_comparator.py [hunk 17]
```diff
@@ -140,9 +151,17 @@ def __init__(self, pairs, threshold=None):
...
+            ).formatted()
 
     def compare(self):
         for pair in self.pairs.values():
             baseline, contender = pair.get("baseline"), pair.get("contender")
-            yield BenchmarkComparator(baseline, contender, self.threshold).compare()
+            yield BenchmarkComparator(
+                baseline,
+                contender,
+                self.threshold,
```
Comment (dianaclarke) [inline conbench/api/_comparator.py]: If you remove the trailing comma after `self.threshold,` there won't be a diff here.

File: conbench/api/compare.py [hunk 18]
```diff
@@ -70,14 +71,24 @@ def get(self, compare_ids):
...
 
         baseline = _compare_entity(baseline_summary)
         contender = _compare_entity(contender_summary)
 
         if raw:
-            return BenchmarkComparator(baseline, contender, threshold).compare()
+            return BenchmarkComparator(
+                baseline,
+                contender,
+                threshold,
```
Comment (dianaclarke) [inline conbench/api/compare.py]: Ditto. Remove trailing comma and re-black.

File: conbench/api/compare.py [hunk 19]
```diff
@@ -70,14 +71,24 @@ def get(self, compare_ids):
...
+            ).compare()
         else:
-            return BenchmarkComparator(baseline, contender, threshold).formatted()
+            return BenchmarkComparator(
+                baseline,
+                contender,
+                threshold,
```
Comment (dianaclarke) [inline conbench/api/compare.py]: Ditto. Remove trailing comma after threshold and re-black.

File: conbench/api/compare.py [hunk 20]
```diff
@@ -132,16 +143,24 @@ def get(self, compare_ids):
...
             set_display_name(summary)
             set_display_batch(summary)
+            set_tags(summary)
             self._add_pair(pairs, summary, "contender")
 
         if raw:
-            result = BenchmarkListComparator(pairs, threshold).compare()
+            result = BenchmarkListComparator(
+                pairs,
+                threshold,
```
Comment (dianaclarke) [inline conbench/api/compare.py]: ditto

File: conbench/api/compare.py [hunk 21]
```diff
@@ -132,16 +143,24 @@ def get(self, compare_ids):
...
+            ).compare()
         else:
-            result = BenchmarkListComparator(pairs, threshold).formatted()
+            result = BenchmarkListComparator(
+                pairs,
+                threshold,
```
Comment (dianaclarke) [inline conbench/api/compare.py]: ditto

**Pull request merged**